keps/sig-apps/85-Graduate-PDB-to-Stable/README.md

```diff
@@ -0,0 +1,319 @@
+has a controller that does not implement scale, and therefore can only be used if the PDB is using `minAvilable` as 
+a number. We also want to introduce conditions on the PDB status object that can signal error situations to
+users and tools. In particular, whenever the failsafe functionality of the disruption controller forces `AllowedDisruptions`
+to be 0, there should be a condition that allow tools to distinguish this situation from `AllowedDisruptions` being 
+0 simply because there are not enough ready pods.
+
+Combined, these should make the behavior of the disruption controller more intuitive, but also provide signals to users
+whenever the configuration of the pdb and/or the targeted workloads are invalid.
+
+#### Address scalability issues with the disruption controller
+
+The disruption controller has some performance issues as reported in https://github.com/kubernetes/kubernetes/issues/92826.
+https://github.com/kubernetes/kubernetes/pull/92827 was merged to remove the 30s resync period which should improve
+performance.
+
+#### Fix handling of empty selector in disruption controller
+
+The disruption controller doesn't handle empty selector correctly https://github.com/kubernetes/kubernetes/issues/95083. 
+We can't fix this directly in the v1beta1 version of PDBs, but we should fix this as part of promoting PDBs to v1.
+
+### API changes
+
+* Add conditions to the status object for PodDisruptionBudget.
+* Change the semantics of an empty selector.
+
+
+### Test Plan
+
+#### Existing Tests
+PodDisruptionBudget currently has tests in various components that use the feature:
+
+* Unit tests for disruption controller
+  - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/disruption/disruption_test.go
+* Integration tests for disruption controller
+  - https://github.com/kubernetes/kubernetes/blob/master/test/integration/disruption/disruption_test.go
+* Kubectl
+  - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/generate/versioned/pdb_test.go
+  tests generation of a PDB objects out of given parameters
+  - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create/create_pdb_test.go
+  tests creation of PDB objects from cmd parameters
+* Scheduler
+  - https://github.com/kubernetes/kubernetes/blob/4b59044b8d2a3502ea490ba2c958008a098511a3/test/integration/scheduler/preemption_test.go#L1053
+  tests effects of PDB on preemption (PDB is honored in a best effort way)
+* Eviction integration tests
+  - https://github.com/kubernetes/kubernetes/blob/master/test/integration/evictions/evictions_test.go test eviction logic and its interactions with PDB.
+* Autoscaler
+  - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/autoscaling/cluster_size_autoscaling.go ensure that Autoscaler respects PDB when draining nodes.
+* Integration tests for disruption controller
+  - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/disruption.go
+
```

Comment:
What new tests are you going to add?
Given that you're planning to address a couple issues, I would expect to see new tests for those.

Reply:
Yeah, there will be tests added for the changes we want to make. I added a section on this.